### DISCUSSION TITLE
[v4] How to add plugins?

### DISCUSSION THREAD

**Discussion by ev3rlost** [tailwindlabs/tailwindcss#13292] [Help] [answered]
How do you add the official plugins with no config file in v4?

**Comment by ev3rlost** [comment]
My bad, I've missed the note saying it isn't available yet in v4 in the roadmap section. I hope this gets added soon!

`
Plugins and custom utilities — we don’t have support for plugins, or for writing custom utilities that automatically work with variants yet. Obviously we will make this work before a stable release.
`

**Reply by fokkerone** [reply]
aha, mouth full of promises. 
the only way is  this hacky @config + @plugin in css directive.. doesnt work with ts and there are alot of plugins / presets that wont work anymore like each thats using breakpoints i.e. 

how often does tw now changed the color type??? 3, 4 times?
no token workflow from figma - json - to reflect all this changes.

I am a bit upset because we need to change presets, themes, a ll color tokens ... literally the comple token workflow in style dictionary... 

alot of work for same-same but different

**Comment by ev3rlost** [comment]
Just a bump post. How's it looking at the plugin department? :) Any news?

**Reply by adamwathan** [reply]
Right now we're working on a few new components for [Catalyst](https://tailwindui.com/templates/catalyst) and releasing Headless UI v2.0 but we'll be back into Tailwind CSS v4 stuff next month and this is one of the top priorities 👍  So expect to see some movement here soon!

**Reply by ev3rlost** [reply]
Go Adam! Thank you for answering! So pumped for v4! :))

**Reply by eliabexp** [reply]
Just a bump post. How's it looking at the plugin department? :) Any news?

**Comment by eloyesp** [comment]
Reading the [CHANGELOG](https://github.com/tailwindlabs/tailwindcss/blob/next/CHANGELOG.md) it seems that plugins will be added using a new css directive `@plugin`.

**Comment by ev3rlost** [comment]
Looking at the[ CHANGELOG](https://github.com/tailwindlabs/tailwindcss/blob/next/CHANGELOG.md), I'm guessing the @plugin directive has been implemented since `4.0.0-alpha.18`?

Does the `tailwindcss/forms` plugin work yet?

How do we implement it? I tried `@plugin "tailwindcss/forms";` but I get `[vite] Internal server error: [postcss] Package subpath './forms' is not defined by "exports"`.

Currently on `[4.0.0-alpha.20]`.

**Reply by adamwathan** [reply, marked as answer]
Should be “@tailwindcss/forms” 👍🏻 don’t forget the “@“ sign!

**Reply by ev3rlost** [reply]
Thank you Adam!

The thing is: I can't npm install the @tailwindcss/forms plugin using the 4.0.0.-alpha.20 release.

I get the following errors:

```
code ERESOLVE
npm error ERESOLVE unable to resolve dependency tree
While resolving: undefined@undefined
npm error Found: tailwindcss@4.0.0-alpha.20
npm error node_modules/tailwindcss
npm error   tailwindcss@"^4.0.0-alpha.20" from the root project
npm error
npm error Could not resolve dependency:
npm error peer tailwindcss@">=3.0.0 || >= 3.0.0-alpha.1" from @tailwindcss/forms@0.5.7
npm error node_modules/@tailwindcss/forms
npm error   dev @tailwindcss/forms@"*" from the root project
Fix the upstream dependency conflict, or retry
npm error this command with --force or --legacy-peer-deps
npm error to accept an incorrect (and potentially broken) dependency resolution.
```

Am I doing something wrong here?

**Reply by kaysef** [reply]
I may be missing this in the changelog somewhere but wondering how to use the plugins with the new `@plugins` directive in v4 and in which file to add these imports. Any example? 

Is it still in `tailwind.config.js` or with v4, that's going away and the plugins are getting added in the `.css` file.

**Reply by ev3rlost** [reply]
Everything resides in the app.css file.

app.css:

```
@import 'tailwindcss';
@plugin '@tailwindcss/forms'; (given you installed the plugin via npm)

@theme {
  --color-primary-500: #f8fafc;
}
```

The problem is you can't install the form plugin as it gives you the above errors.

**Reply by philipp-spiess** [reply]
@ev3rlost You can try to install it with `--force` (which ignores the upstream version conflict):

```
npm install @tailwindcss/forms --force
```

We'll get the peer dependency version fixed though, thanks!

**Reply by philipp-spiess** [reply]
We have updated both the `@tailwindcss/forms` and `@tailwindcss/typography` packages now. It now works without the `--force` 🙂

**Reply by dofirfauzi** [reply]
How to add options plugin like in config file ?
plugins: [
  require("@tailwindcss/forms")({
    strategy: 'base', // only generate global styles
    strategy: 'class', // only generate classes
  }),
],

@plugin "@tailwindcss/forms"; how to add the option ??
@adamwathan

**Reply by scott1702** [reply]
This can be done in CSS like this
```css
@plugin "@tailwindcss/forms" {
  strategy: 'class';
}
```

**Comment by ev3rlost** [comment]
Thank you philipp! Adding the --force flag worked! Glad to see it being fixed.

**Comment by Stephan-MC** [comment]
What about packages with multiple plugin functions that I may need to use only one of them.

**Reply by philipp-spiess** [reply]
For now you have to create a custom `.js` file that imports the package and only has one default export, so something like this should work:

```js
/* my-plugin-x.js */
import { PluginX } from 'my-tailwind-plugins';

export default PluginX;
```

```css
/* root.css */
@plugin './my-plugin-x.js';
```

**Reply by Stephan-MC** [reply]
ok  thanks. but might become tedious if i have multiple plugins to work with. :+1:

**Reply by SantaClaas** [reply]
I was wondering the same but for my plugin calls a function to configure the returned plugin function. Thanks

**Comment by HarshArora-1205** [comment]
@plugin 'tailwindcss-animate';
@plugin '@tailwindcss/forms';
@plugin '@tailwindcss/typography';

When I include plugins, I receive a warning: Unknown at rule @plugin css(unknownAtRules)

But I am able to use them. 
tailwind version - 4.0.3.

**Reply by wongjn** [reply]
You can install and set up the Tailwind Intellisense plugin if you want to get rid of the warnings in VSCode.

**Reply by HarshArora-1205** [reply]
> You can install and set up the Tailwind Intellisense plugin if you want to get rid of the warnings in VSCode.

I already have intellisense installed, still I receive those warnings.

**Reply by Stephan-MC** [reply]
you'll need to disable css linting in your vs code.

this error is shown because the built in css lsp in your vs code doesn't recognize ```@plugin``` as a valid css at-rule.

**Reply by agoodkind** [reply]
> you'll need to disable css linting in your vs code.
> 
> this error is shown because the built in css lsp in your vs code doesn't recognize `@plugin` as a valid css at-rule.

But what if CSS linting is desired?

**Reply by maribox** [reply]
You need to select "Tailwind CSS" on the bottom right in VS Code (when the Tailwind CSS IntelliSense extension is installed)

**Reply by ujjaldey** [reply]
> You need to select "Tailwind CSS" on the bottom right in VS Code (when the Tailwind CSS IntelliSense extension is installed)

Thanks! It worked!

![image](https://github.com/user-attachments/assets/6e389722-fc9c-4e4d-a40f-0fa57d155ac2)

**Comment by amantiwari8861** [comment]
how to install rippleui(https://www.ripple-ui.com/docs/get-started/installation) plugin in tailwind 4.0

**Reply by obouchari** [reply]
Same as above:
```
@plugin "rippleui";
```

**Reply by abodnar-tis-tech** [reply]
@obouchari  @amantiwari8861  Did you successfully install the RippleUI plugin? I tried to use it, but when starting my app I received the following _error:_
```
⨯ ./src/styles/globals.css.webpack[javascript/auto]!=!./node_modules/next/dist/build/webpack/loaders/css-loader/src/index.js??ruleSet[1].rules[13].oneOf[10].use[2]!./node_modules/next/dist/build/webpack/loaders/postcss-loader/src/index.js??ruleSet[1].rules[13].oneOf[10].use[3]!./src/styles/globals.css
Error: `addUtilities({ '@supports (-moz-appearance:none)' : … })` defines an invalid utility selector. Utilities must be a single class name and start with a lowercase letter, eg. `.scrollbar-none`
```

Also my previous tw config: 

```typescript
module.exports = {
 plugins: [require("rippleui"), require("tailwindcss-animate")],
  /** @type {import('rippleui').Config} */
  rippleui: {
    removeThemes: ["dark"],
  },
}

``` 




Please share your experience with this.

**Comment by zenovak** [comment]
Hello Everyone, what about custom plugins that declare a new variant? Im trying to update the following code:
```js
module.exports = {
 plugins: [
    plugin(function({ addVariant }) {
      addVariant('slider-thumb', ['&::-webkit-slider-thumb', '&::slider-thumb', '&::-moz-range-thumb']);
      addVariant('slider-track', ['&::-webkit-slider-runnable-track', '&::-moz-range-track']);
    }),
  ],
}
```
  How do we do this in v4.0? Thanks.

**Reply by katerlouis** [reply]
I've been searching for half an hour now already and am baffled that no one has asked that before. @zenovak have you found a solution? I'm really confused that programmatic variant generation does not seem to be a thing anymore. I have build a system with rather complex custom plugins over the years and while technically possible, it would be a gigantic pain to "write out" the hundreds of utilities using `@plugin` inside a css file oO

**Reply by tom-wenger** [reply]
Any updates to self made plugins? I had a plugin who generated utitlity classes with different breakpoints inside.
like this: 
```
const newUtilities = [
'.c-p-xsmall': {
    padding: '12px',
    '@screen 2xl': { padding: '18px' },
    '@screen 3xl': { padding: '24px' }
  },
  '.c-p-small': {
    padding: '7px',
    '@screen md': { padding: '12px' },
    '@screen lg': { padding: '14px' },
    '@screen xl': { padding: '21px' },
    '@screen 3xl': { padding: '28px' }
  }
]
addUtilities(newUtilities)
```
but it seems like @screen doesnt get processed by tailwind as it did before. and when i try generating @media entries i get this error: `addUtilities({ '@media (min-width: 1400px)' : … }) defines an invalid utility selector. Utilities must be a single class name and start with a lowercase letter, eg. .scrollbar-none.`

**Reply by yannickoo** [reply]
Hey @tom-wenger, I hope you could resolve your plugin in the meantime. I just had to fix the same issue and have realized that `@screen` is not valid anymore and you would need to write a media query instead.

Instead of writing this:

```
addUtilities([
  '.c-p-xsmall': {
    padding: '12px',
    '@screen 2xl': { padding: '18px' },
    '@screen 3xl': { padding: '24px' }
  },
  '.c-p-small': {
    padding: '7px',
    '@screen md': { padding: '12px' },
    '@screen lg': { padding: '14px' },
    '@screen xl': { padding: '21px' },
    '@screen 3xl': { padding: '28px' }
  }
])
```

you would need to read the breakpoints from your config and do this here instead:

```
const screens = theme('screens', {})
addUtilities([
  '.c-p-xsmall': {
    padding: '12px',
    [`@media (min-width: ${screens.2xl})`]: { padding: '18px' },
    [`@media (min-width: ${screens.3xl})`]: { padding: '24px' }
  },
  '.c-p-small': {
    padding: '7px',
    [`@media (min-width: ${screens.md})`]: { padding: '12px' },
    [`@media (min-width: ${screens.lg})`]: { padding: '14px' },
    [`@media (min-width: ${screens.xl})`]: { padding: '21px' },
    [`@media (min-width: ${screens.2xl})`]: { padding: '28px' }
  }
])
```

_Quick tip from my side: `7px` or `21px` look a bit weird and maybe it's better to switch to rem-based values 🤓_

**Comment by tecoad** [comment]
@adamwathan i found [this](https://github.com/tailwindlabs/tailwindcss/blob/63b9be94838d68c12e34014b4cf9dc2927c24e15/packages/tailwindcss/src/plugin.ts#L11) createPlugin and exported legacy types, but cannot find how plugins can be created on v4 anymore.

The syntax i was adopting on v3 was:

```
plugin((api: PluginAPI) => {
			const { addUtilities } = api

			const lineCounts = [1, 2, 3] // Define the number of grid lines you need
			const lineStyles = ['solid', 'dashed'] // Define the line styles
```


Could you shed some light on the issue? Will this never work in v4 or is implementation different?
Due to this seems i can not just migrate yet.

**Reply by SantaClaas** [reply]
I'm not sure what you exactly mean.
If it is the plugin function you are looking for then it can be imported from `tailwindcss/plugin` in v4.
As for the types. They seem to be not exported right now but a PR to fix it seems to be already merged. See https://github.com/tailwindlabs/tailwindcss/pull/16505#issue-2851110078 so probably in the next version.
Hope that helps.

**Reply by rasulovdev** [reply]
In v4 docs, `@plugin` directive is described under "[Compatibility](https://tailwindcss.com/docs/functions-and-directives#compatibility)" section, which states "The following directives and functions exist solely for compatibility with Tailwind CSS v3.x.". Moreover, `@plugin` directive description says "Use the @plugin directive to load a **_legacy JavaScript-based_** plugin". 
That is leading me toward the same question OP has: does that all mean there's a "modern" way of composing plugins?

**Reply by eloyesp** [reply]
As all the configuration is done within `css` files, and `@import` is a native feature on tailwind 4.

So a **modern plugin** is a `plugin.css` file and to use it just `@import "plugin.css"`.

**Reply by kpietraszko** [reply]
The "modern" way of creating plugins (or rather utilities) is documented here: https://tailwindcss.com/docs/adding-custom-styles#adding-custom-utilities

If you want a more complete example, I posted one here: https://github.com/tailwindlabs/tailwindcss/discussions/5541#discussioncomment-12394063

**Comment by rozsazoltan** [comment]
Related:
* https://github.com/tailwindlabs/tailwindcss/discussions/15997

**Comment by oyzamil** [comment]
## Solution for custom-made plugins 💡

To recap, for **custom-made plugins**, you no longer need to modify `tailwind.config.js`.

Create your plugin separately and import it into your `tailwind.css` file.

### Sample Plugin ⚡
Here is a quick sample plugin for a custom color palette.

**tailwind-color.ts**
```
import Color from "color";
import plugin from "tailwindcss/plugin";

// helper to convert to alpha-capable rgb()
function toRgbWithAlpha(hex: string) {
  const [r, g, b] = Color(hex).rgb().array();
  return `rgb(${r} ${g} ${b} / <alpha-value>)`;
}

export default plugin.withOptions(
  () => {
    // no custom utilities needed, just theme extension
    return function () {};
  },
  () => {
    const base = Color('#fc801c');

    const shades = {
      50: base.lighten(0.9).hex(),
      100: base.lighten(0.7).hex(),
      200: base.lighten(0.5).hex(),
      300: base.lighten(0.3).hex(),
      400: base.lighten(0.1).hex(),
      500: base.hex(),
      600: base.darken(0.1).hex(),
      700: base.darken(0.3).hex(),
      800: base.darken(0.5).hex(),
      900: base.darken(0.7).hex(),
    };

    // convert to alpha-aware form
    const alphaShades = Object.fromEntries(
      Object.entries(shades).map(([k, v]) => [k, toRgbWithAlpha(v)]),
    );

    return {
      theme: {
        extend: {
          colors: {
            app: alphaShades,
          },
        },
      },
    };
  },
);

```

**tailwind.css**
```
@import "tailwindcss";
@plugin './tailwind-color.ts';
@layer base {}
```
Add as many plugins as you want. 🎊

Regards, [Muzammil](https://softwebtuts.com)

**Reply by martinsik** [reply]
Where does `plugin.withOptions` come from? There's not a single mention of this in the official documentation at https://tailwindcss.com/docs

**Reply by crswll** [reply]
https://v3.tailwindcss.com/docs/plugins#exposing-options

**Comment by agoodkind** [comment]
What about v4, I believe that was the original reason for this thread (it
says it in the title no less)

**Reply by oyzamil** [reply]
Everything else stays the same as mentioned above; instead of importing the plugin in the `tailwind.css` file, you need to import it in `tailwind.config.ts`.

```
/** @type {import('tailwindcss').Config} */
module.exports = {
  theme: {
    // ...
  },
  plugins: [
    require('./tailwind-color.ts')
  ],
}
```

**Reply by rozsazoltan** [reply]
```css
@import "tailwindcss";
@plugin "./path/to/tailwind-color.ts";
```
